docs/proposals/20210427-unitedtopology.md

```diff
@@ -0,0 +1,256 @@
+---
+title: UnitedTopology
+authors:
+- "@FillZpp"
+reviewers:
+- "@Fei-Guo"
+- "@furykerry"
+creation-date: 2021-04-27
+last-updated: 2021-04-27
+status: implementable
+---
+
+# UnitedTopology
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a proposal and for highlighting
+any additional information provided beyond the standard proposal template.
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc) a table of contents from markdown are available.
+
+
+## Motivation
+
+Define a CRD named UnitedTopology, which can support configuration and management for elastic and multi-domain applications,
+even more powerful and flexible than [UnitedDeployment](https://openkruise.io/en-us/docs/uniteddeployment.html) in some ways.
+
+Not like the workload type, UnitedTopology can be used with those stateless workloads, such as CloneSet, Deployment and ReplicaSet.
+
+So users needn't learn the usage of a new workload and how to deploy and update their application by it.
+They only have to create UnitedTopology CRs related to their existing workloads. Everything is just great!
+
+## Proposal
+
+UnitedTopology definition looks like this:
+
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: UnitedTopology
+metadata:
+  namespace: ...
+  name: ...
+  # ...
+spec:
+  targetRef:
+    apiVersion: apps/v1 | apps.kruise.io/v1alpha1
+    kind: Deployment | CloneSet
+    name: workload-xxx
+  subsets:
+  - name: subset-a
+    nodeSelectorTerm:
+      matchExpressions:
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values:
+        - zone-a
+    maxReplicas: 3
+    patch:
+      metadata:
+        labels:
+          xxx-specific-label: xxx
+  - name: subset-b
+    nodeSelectorTerm:
+      matchExpressions:
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values:
+        - zone-b
+  scheduleStrategy:
+    type: Adaptive | Fixed
+    adaptive:
+      disableSimulationSchedule: false
+      rescheduleCriticalSeconds: 30
+status:
+  observedGeneration: 12
+  subsetStatuses:
+  - name: subset-a
+    missingReplicas: 2
+    creating:
+      pod-xxx: "2021-04-27T08:34:11Z"
+    deleting:
+      pod-yyy: "2021-04-27T08:34:11Z"
+  - name: subset-b
+    missingReschedule: 1
+```
+
+**spec**:
+- `targetRef`: reference to a specific workload, it supports `Deployment` in `apps` group and `CloneSet` in `apps.kruise.io` group
+- `subsets`: list of subsets which defines different topologies
```

Comment:
comments about the order of topology which may influence creation and deletion order